# PWD: frameworks/base

diff --git a/packages/SystemUI/src/com/android/systemui/qs/external/CustomTile.java b/packages/SystemUI/src/com/android/systemui/qs/external/CustomTile.java
index 30e0a766de3..cfefd486cb7 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/external/CustomTile.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/external/CustomTile.java
@@ -326,7 +326,7 @@ public class CustomTile extends QSTileImpl<State> implements TileChangeListener
         } catch (Exception e) {
             Log.w(TAG, "Invalid icon, forcing into unavailable state");
             state.state = Tile.STATE_UNAVAILABLE;
-            drawable = mDefaultIcon.loadDrawable(mUserContext);
+            drawable = (mDefaultIcon != null) ? mDefaultIcon.loadDrawable(mUserContext) : null;
         }
 
         final Drawable drawableF = drawable;
