# PWD: external/selinux

diff --git a/prebuilts/bin/audit2allow b/prebuilts/bin/audit2allow
index 0d96ca60..8be60117 100755
--- a/prebuilts/bin/audit2allow
+++ b/prebuilts/bin/audit2allow
@@ -16,7 +16,7 @@ unamestr=`uname`
 if [ "$unamestr" = "Linux" -o "$unamestr" = "linux" ]; then
 	export LD_PRELOAD=$ANDROID_HOST_OUT/lib64/libselinux.so
 	export PYTHONPATH=$ANDROID_BUILD_TOP/prebuilts/python/linux-x86/2.7.5/lib/python2.7/site-packages
-	python $ANDROID_BUILD_TOP/external/selinux/python/audit2allow/audit2allow "$@"
+	python2 $ANDROID_BUILD_TOP/external/selinux/python/audit2allow/audit2allow "$@"
 else
 	echo "audit2allow is only supported on linux"
 	exit 1
diff --git a/python/audit2allow/audit2allow b/python/audit2allow/audit2allow
index 09b06f66..57a4c805 100755
--- a/python/audit2allow/audit2allow
+++ b/python/audit2allow/audit2allow
@@ -1,4 +1,4 @@
-#!/usr/bin/python3 -Es
+#!/usr/bin/python2 -Es
 # Authors: Karl MacMillan <kmacmillan@mentalrootkit.com>
 # Authors: Dan Walsh <dwalsh@redhat.com>
 #
