diff --git a/EnumType.cpp b/EnumType.cpp
index 32682645..9f808eaa 100644
--- a/EnumType.cpp
+++ b/EnumType.cpp
@@ -289,7 +289,12 @@ void EnumType::emitIteratorDeclaration(Formatter& out) const {
         elementCount += type->mValues.size();
     }
 
-    out << "template<> constexpr std::array<" << getCppStackType() << ", " << elementCount
+    // TODO(pcc): Remove the pragmas once all users of the hidl headers have
+    // been moved to C++17.
+    out << "#pragma clang diagnostic push\n";
+    out << "#pragma clang diagnostic ignored \"-Wc++17-extensions\"\n";
+
+    out << "template<> inline constexpr std::array<" << getCppStackType() << ", " << elementCount
         << "> hidl_enum_values<" << getCppStackType() << "> = ";
     out.block([&] {
         auto enumerators = typeChain();
@@ -300,6 +305,8 @@ void EnumType::emitIteratorDeclaration(Formatter& out) const {
             }
         }
     }) << ";\n";
+
+    out << "#pragma clang diagnostic pop\n";
 }
 
 void EnumType::emitEnumBitwiseOperator(
